io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/writer/internal/StreamGmlWriter.java

```diff
@@ -302,53 +354,233 @@ protected IOReport execute(ProgressIndicator progress, IOReporter reporter)
+				PARTITION_BY_EXTENT_MODE_DATASET);
+
+		final SubtaskProgressIndicator qtProgress = new SubtaskProgressIndicator(progress) {
+
+			@Override
+			protected String getCombinedTaskName(String taskName, String subtaskName) {
+				return taskName + " (" + subtaskName + ")";
+			}
+
+		};
 
-			final Set<TypeDefinition> types = new HashSet<>();
+		// Map for instances that either contain no or multiple geometries
+		Map<String, InstanceReference> unhandledInstances = new HashMap<>();
+
+		QuadtreeBuilder<Point, InstanceReference> builder = new QuadtreeBuilder<>();
+		try (ResourceIterator<Instance> it = getInstances().iterator()) {
+			qtProgress.begin("Build quadtree", getInstances().size());
 
-			// Map GML IDs to features types and collect types
 			final XMLInspector gadget = new XMLInspector();
-			final Map<String, TypeDefinition> idToTypeMapping = new HashMap<>();
-			try (ResourceIterator<Instance> it = getInstances().iterator()) {
-				while (it.hasNext()) {
-					Instance inst = it.next();
-					types.add(inst.getDefinition());
-					idToTypeMapping.put(gadget.getIdentity(inst), inst.getDefinition());
+			int i = 0;
+			while (it.hasNext()) {
+				Instance inst = it.next();
+				InstanceReference instRef = getInstances().getReference(inst);
+
+				InstanceTraverser traverser = new DepthFirstInstanceTraverser();
+				GeometryFinder finder = new GeometryFinder(getTargetCRS());
+				traverser.traverse(inst, finder);
+				List<GeometryProperty<?>> geoms = finder.getGeometries();
+				if (geoms.isEmpty() || geoms.size() > 1) {
+					unhandledInstances.put(gadget.getIdentity(inst), instRef);
+				}
+				else {
+					GeometryProperty<?> geomProperty = geoms.get(0);
+
+					Geometry geom = geomProperty.getGeometry();
+
+					Point centroid;
+					switch (mode) {
+					case PARTITION_BY_EXTENT_MODE_WORLD:
+						CoordinateReferenceSystem sourceCrs = geomProperty.getCRSDefinition()
+								.getCRS();
+						CodeDefinition wgs84 = new CodeDefinition("EPSG:4326");
+						try {
+							MathTransform toWgs84 = CRS.findMathTransform(sourceCrs,
+									wgs84.getCRS());
+							Geometry geomWgs84 = JTS.transform(geom, toWgs84);
+							centroid = geomWgs84.getCentroid();
+						} catch (FactoryException | MismatchedDimensionException
+								| TransformException e) {
+							log.error("Unable to transform geometry to WGS 84", e);
+							throw new IllegalStateException(e.getMessage(), e);
+						}
+						break;
+					case PARTITION_BY_EXTENT_MODE_DATASET:
+						// fall through to default
+					default:
+						centroid = geom.getCentroid();
+					}
+
+					builder.add(centroid,
+							new IdentifiableInstanceReference(instRef, gadget.getIdentity(inst)));
+				}
+
+				qtProgress.advance(1);
+				if (++i % 100 == 0) {
+					qtProgress.setCurrentTask(
+							MessageFormat.format("Adding geometries ({0} instances processed)", i));
 				}
 			}
+			qtProgress.end();
+			qtProgress.begin("Building quadtree", ProgressIndicator.UNKNOWN);
```

Comment:
There does not seem to be a semantical difference in the task name before ("Build quadtree") and this one.